src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/AntiforgeryValidationFailedResult.cs

```diff
@@ -0,0 +1,35 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Antiforgery;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Mvc.ViewFeatures.Internal
+{
+    /// <summary>
+    /// Antiforgery validation failed action result, that defauls to the
+    /// behavior of <see cref="BadRequestResult"/> unless there is a custom
+    /// handler configured in the <see cref="AntiforgeryOptions"/>.
+    /// </summary>
+    public class AntiforgeryValidationFailedResult : BadRequestResult
+    {
+        public override Task ExecuteResultAsync(ActionContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var options = context.HttpContext.RequestServices.GetService<IOptions<AntiforgeryOptions>>().Value;
```

Comment:
Why not just write a result filter to handle this? Moreover the layering is weird with this approach, the result is an MVC concern, why would you expect the handler to be defined in Antiforgery?

Reply:
Oh. I was under the impression that since the antiforgery is running as an authorization filter that’s triggering the short-circuiting behavior of the MVC filter pipeline, that there was no way to have a result filter run afterwards. While that’s true for normal result filters, `AlwaysRunResultFilters` do work there – I did not know that those existed! That makes this a lot easier of course.

As for the mixed concerns, I had the handler defined in the `AntiforgeryOptions` simply to avoid introducing a new MVC-specific antiforgery options type just for configuring the handler.

But with `IAsyncAlwaysRunResultFilter` working for this purpose, this gets a lot simpler, and I won’t need to touch those options. I’ll update the PR for this. – Thanks for the idea!